### PR TITLE
[RHELC-16] Do not disable RHSM repos before pkg backup

### PR DESCRIPTION
**Pull request by hosekadam**
There is a problem with backing up the packages on systems which use satellite server. During the conversion is system unregistered from previous subscription - in that step we lose access to the CentOS repositories. And then we need to backup packages, which could come from CentOS. 

So the step with backup and remove the package is moved before the unregistration. With that show up another problem - the backed up and then removed package remove `os-release` file too. The file is needed for subscribing the system, so it is restored from backup and then removed (if it won't be removed, the file would stay there after conversion as  `os-release.rpmorig`)

Jira Issue: [RHELC-16](https://issues.redhat.com/browse/RHELC-16)

Checklist
- [x] PR meets acceptance criteria specified in the Jira issue
- [x] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending`

### CODE REVIEW

File: convert2rhel/subscription.py
Comment (SpyTec) [inline convert2rhel/subscription.py:185]: Add a comment why we need to do this. If I were to look at this code after a while I wouldn't know why we're doing it without going through an issue

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:209]: Add a comment about needing to remove the repofile packages before disabling repos

File: convert2rhel/backup.py
Comment (SpyTec) [inline convert2rhel/backup.py:128]: Probably an info log would be helpful here
```suggestion
        if rollback:
            loggerinst.task("Rollback: Restoring %s from backup" % self.filepath)
        else:
            loggerinst.info("Restoring %s from backup" % self.filepath)
```

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/backup.py
Comment (hosekadam) [reply convert2rhel/backup.py:128]: @SpyTec What do you think about changing the other logs [1](https://github.com/oamg/convert2rhel/pull/627/files#diff-8f4622020cde9a253bc82f69bf2b88910bcc5a84943db3c7d3a435becedfd4d5R145), [2](https://github.com/oamg/convert2rhel/pull/627/files#diff-8f4622020cde9a253bc82f69bf2b88910bcc5a84943db3c7d3a435becedfd4d5R151) and [3](https://github.com/oamg/convert2rhel/pull/627/files#diff-8f4622020cde9a253bc82f69bf2b88910bcc5a84943db3c7d3a435becedfd4d5R153) to `info` too? My idea of using the `debug` was to not complicate the output for user, but now I'm thinking if it could be somehow useful if the backup would be there.

Comment (SpyTec) [reply convert2rhel/backup.py:292]: @hosekadam The links don't work for me but I assume you mean the other sections with restore complete etc. For those kinds of logs I would say this:
- if it is important to message the user about a completion, it should be `info` for sure. That's why having it when rolling back is good
- if the message is part of any task and isn't critical, the task completion should be a debug as it otherwise is unnecessary clutter

The reason I wanted to change this is that it relates to a task's start. As customers or problem solvers (like us or our support engineers) would like to know what happened to lead to an issue. If we back something up and something fails, without logs we'd make the assumption it was the previous task. Or be let in the dark of what happened without digging into the code itself. As for using `task` vs `info`, when rolling back it is a crucial step to restore backups, so a task make sense there. Though if it's within the intended workflow of converting a system it would be confusing and thus an `info` log is good enough

Comment (hosekadam) [reply convert2rhel/backup.py:292]: @SpyTec Ok, so we can leave the rest as `debug`. And thanks for the clarification :+1:

File: convert2rhel/backup.py
Comment (r0x0d) [inline convert2rhel/backup.py:332]: Typo 
```suggestion
            loggerinst.debug("Couldn't remove restored file %s" % self.filepath)
```

File: convert2rhel/backup.py
Comment (r0x0d) [inline convert2rhel/backup.py:331]: I'm wondering why we need to have the `TypeError` here. Could you elaborate on why it is needed to catch this exception too? 

If it is something way off that you have experienced in your tests, then it might be good to add a comment explaining the reason to it.

Comment (hosekadam) [reply convert2rhel/backup.py:331]: @r0x0d The `TypeError` exception would be raised, if the `filepath` would somehow contain `None`. Which I think couldn't happen, but if yes, I've choose that just saying `Couldn't remove restored file` is better than crashing on this.

Comment (r0x0d) [reply convert2rhel/backup.py:331]: Right. Makes sense from this point of view, but since we construct the `RestorableFile` object from hard-coded strings in our code, then I don't think we need the `TypeError` catch here, as we will always pass an `filepath` to the constructor of the class.

Catching the `TypeError` in this function, for me, seems unnecessary, as it would be better if we catch that type of mistaken in the class constructor, for example: 

```python
class RestorableFile(object):
    def __init__(self, filepath):
        if not isinstance(filepath, str):
            raise TypeError("filepath must be an instance of str.")
            
        self.filepath = filepath
```

This way we could prevent working with an empty `filepath` in other methods inside the class.

But for now, I would say this is not strictly necessary, as we don't have any case in our code base that could pass an empty `filepath` to this class constructor.

Comment (hosekadam) [reply convert2rhel/backup.py:331]: Ok, so I'll remove that.

**Review by SpyTec — APPROVED** [review]

**Comment by hosekadam** [comment]
Heh, I used probably a wrong command and I see there is a bit mess...

**Comment by hosekadam** [comment]
There are few things left that needs to be done to get the integration test running properly:
1. Merge this [merge request](https://gitlab.cee.redhat.com/oamg/convert2rhel/convert2rhel-secrets/-/merge_requests/9)
2. In the satellite repo for CentOS7 is the old version of kernel, so we need to wait to sync the repositories

Otherwise the test works, it's just changed the test for offline conversion using satellite to use the new satellite dogfood server

**Comment by r0x0d** [comment]
> There are few things left that needs to be done to get the integration test running properly:
> 
>     1. Merge this [merge request](https://gitlab.cee.redhat.com/oamg/convert2rhel/convert2rhel-secrets/-/merge_requests/9)
> 
>     2. In the satellite repo for CentOS7 is the old version of kernel, so we need to wait to sync the repositories
> 
> 
> Otherwise the test works just for the CentOS 7 so I think I should limit that just for this system and not to run it with others. The second variant is to ask satellite guys to provide servers that we need for another systems. What do you think @SpyTec or @r0x0d?

@hosekadam ping @kokesak to merge your PR. I don't have write access to that repository. 

I don't know how the satellite guys would react if we ask them for more servers for the other distros we use… Furthermore, I don't know if @SpyTec or @bocekm knows how that works, if so, can any of you help us here? 

As far as limiting the test to CentOS 7, I think it's a good idea if your test is only related to C

File: tests/integration/conftest.py
Comment (kokesak) [inline tests/integration/conftest.py:29]: We are actually trying to use the new satellite URL (https://issues.redhat.com/browse/RHELC-732). The dogfood is going to be removed.

**Comment by kokesak** [comment]
The test was running for all distros so I would keep that.

IIUC, it's the offline-conversion test in which we currently use the env variables to omit bug this PR fixes. However, we need centos and oracle content in the satellite server, currently, we have just centos7 right?. Is this fix going to be in the next release?

I would rather have it tested on all distros before merging.

File: tests/integration/conftest.py
Comment (hosekadam) [reply tests/integration/conftest.py:29]: Yeah, I see. But if you register the system using subscription-manager to get the CentOS 7 satellite server, there are still urls to the dogfood in `/etc/yum.repos.d/redhat.repo`. Maybe adding both of them, so once it would be finally changed, nothing breaks and just be needed to remove the dogfood url from `/etc/dnsmasq.conf`?

**Comment by hosekadam** [comment]
@kokesak yes, I see [there](https://gitlab.cee.redhat.com/satellite/satellite-ansible-setup/-/merge_requests/12/diffs?commit_id=dca405d5de75b63a753a1b871045b20ff012208d) that there is available satellite just for CentOS 7 (and Centos 8 Stream, but it's useless for us I think). 

And this PR should be in the next release (from Jira).

File: convert2rhel/subscription.py
Comment (abadger) [inline convert2rhel/subscription.py:191]: This part feels a little unclean but we can fix it in the next release.  I've written up what's unclean about it and how we can fix it later here: https://issues.redhat.com/browse/RHELC-809

File: convert2rhel/subscription.py
Comment (abadger) [inline convert2rhel/subscription.py:185]:
```suggestion
            # The file /etc/os-release is needed for subscribing the system and is being removed with
            # <system-name>-release package in one of the previous steps:
            # Fixes RHELC-16
            #
            # This code should be cleaned up in the future.  See  https://issues.redhat.com/browse/RHELC-809
            # for what should be done.
```

**Comment by kokesak** [comment]
On centos7 it seems that the satellite repos do not have the kernel we update to during the update system phase:
```
22:17:29         out: CRITICAL - The version of the loaded kernel is different from the latest version in the enabled system repositories.
22:17:29         out:  Latest kernel version available in Satellite_Engineering_CentOS_7_Updates_x86_64: 3.10.0-1160.76.1.el7
22:17:29         out:  Loaded kernel version: 3.10.0-1160.80.1.el7
22:17:29         out: 
22:17:29         out: To proceed with the conversion, update the kernel version by executing the following step:
22:17:29         out: 
22:17:29         out: 1. yum install kernel-3.10.0-1160.76.1.el7 -y
22:17:29         out: 2. reboot
22:17:29         out: [11/18/2022 17:17:29] DEBUG - Traceback (most recent call last):
22:17:29         out:   File "/usr/lib/python2.7/site-packages/convert2rhel/main.py", line 101, in main
22:17:29         out:     checks.perform_system_checks()
22:17:29         out:   File "/usr/lib/python2.7/site-packages/convert2rhel/checks.py", line 127, in perform_system_checks
22:17:29         out:     is_loaded_kernel_latest()
22:17:29         out:   File "/usr/lib/python2.7/site-packages/convert2rhel/checks.py", line 780, in is_loaded_kernel_latest
22:17:29         out:     "2. reboot" % (repos_message, repoid, latest_kernel, loaded_kernel, package_to_check, latest_kernel)
22:17:29         out:   File "/usr/lib/python2.7/site-packages/convert2rhel/logger.py", line 195, in _critical
22:17:29         out:     sys.exit(msg)
22:17:29         out: SystemExit: The version of the loaded kernel is different from the latest version in the enabled system repositories.
22:17:29         out:  Latest kernel version available in Satellite_Engineering_CentOS_7_Updates_x86_64: 3.10.0-1160.76.1.el7
22:17:29         out:  Loaded kernel version: 3.10.0-1160.80.1.el7
22:17:29         out: 
22:17:29         out: To proceed with the conversion, update the kernel version by executing the following step:
22:17:29         out: 
22:17:29         out: 1. yum install kernel-3.10.0-1160.76.1.el7 -y
22:17:29         out: 2. reboot
```

**Comment by hosekadam** [comment]
@kokesak Yes, I've mentioned this problem in the [comment](https://github.com/oamg/convert2rhel/pull/627#issuecomment-1314111669) :D Do you know somebody who we can contact to help us with that? I thought, they (who prepared the satellite server) have some automated way to stay updated with the original repositories. 

When I was testing that, I used the older version of vagrant box and removed the update step from ansible setup.

**Comment by kokesak** [comment]
Hmm, seem that passing the the env var `CONVERT2RHEL_UNSUPPORTED_SKIP_KERNEL_CURRENCY_CHECK` did not work. I think I will need to make some other workaround, though it's weird that this envar is ommited

File: plans/tier1.fmf
Comment (danmyway) [inline plans/tier1.fmf:244]: Do we need to reboot?
Wouldn't `grubby --set-default-kernel` be sufficient?
I would probably add some comment about the kernel workaround being temporary.
Or a `#TODO` to remove it, when the repository gets synced over at Satellite.
But I guess for the sake of saving some time, we can also keep it as is.

File: tests/integration/tier1/convert-offline-systems/main.fmf
Comment (danmyway) [inline tests/integration/tier1/convert-offline-systems/main.fmf:3]: Summary should be shorter, 50 characters if possible.
Description can be longer.
But we can fix this whenever.

**Review by danmyway — APPROVED** [review]
Just a couple of minor changes, we can sort this out with another PR.

File: plans/tier1.fmf
Comment (kokesak) [reply plans/tier1.fmf:244]: The reboot would be needed I believe as the c2r check is done against the running kernel which changes only after reboot.

File: tests/integration/tier1/convert-offline-systems/main.fmf
Comment (kokesak) [reply tests/integration/tier1/convert-offline-systems/main.fmf:3]: I created a Jira issue for this so we can tackle all the missing stuff in future fix and not block current PR.
Link to the Jira: https://issues.redhat.com/browse/RHELC-824

File: convert2rhel/backup.py
Comment (r0x0d) [inline convert2rhel/backup.py:324]: This message here is not making too much sense at the moment, I would say... 

If it take as an example the same condition above, we are issuing an `loggerinst.task()` when we have a rollback, and a `loggerinst.info()` when we are not rollbacking. 

I would suggest following the same pattern as above, but this is not a blocker. 

```suggestion
        if rollback:
            loggerinst.task("Rollback: File %s restored." % self.filepath)
        else:
            loggerinst.info("File %s restored." % self.filepath)
```

**Review by r0x0d — APPROVED** [review]

File: convert2rhel/backup.py
Comment (SpyTec) [reply convert2rhel/backup.py:324]: I'd disagree. If it's a rollback we'd want to have it as a step, so it would be 
```
[Task] Rollback: Restoring %s from backup
[Info] Rollback succeeded
```

Since at that step we care about the rollback and not about a generic backup we should make a not of it. If it's not a rollback it should be part of the logs as normal, as it's part of the overall task it's part of
```
[Info] Rollback: Restoring %s from backup
[Debug] Rollback succeeded
```

I wouldn't change it to two tasks as that separates it and creates a division that can be understood in the logs as being part of two functions

Comment (bocekm) [reply convert2rhel/backup.py:324]: The original non-task way is IMHO correct. We want just the first line to have the TASK prefix, and the "confirmation" message to be just of an info level below it. Example of what we have now:
```
10:34:02                 out: [11/29/2022 05:34:02] TASK - [Rollback: Restoring /etc/system-release from backup] **************
10:34:02                 out: File /etc/system-release restored.
```

With your proposal it would be:
```
10:34:02                 out: [11/29/2022 05:34:02] TASK - [Rollback: Restoring /etc/system-release from backup] **************
10:34:02                 out: [11/29/2022 05:34:02] TASK - [Rollback: File /etc/system-release restored.] **************
```

Comment (r0x0d) [reply convert2rhel/backup.py:324]: Makes sense to me. Thanks for the explanation

**Pull request merged**
